frontend/model/database.js

```diff
@@ -164,19 +164,76 @@ const files = localforage.createInstance({
   storeName: 'Files'
 })
 
+const maxFileEntries = 100
+
 sbp('sbp/selectors/register', {
-  'gi.db/files/save': function (url: string, blob: Blob): Promise<*> {
-    return files.setItem(url, blob).then(v => {
-      console.log('successfully saved:', url)
-    }).catch(e => {
-      console.error('error saving:', url, e)
+  'gi.db/files/save': async function (cacheKey: string, blob: Blob): Promise<*> {
+    const keys = await files.getItem('keys') ?? []
+
+    // We need to perform several operations in the DB, which includes
+    // the operation requested (i.e., saving a file) and updating the `keys`
+    // entry for caching, as well as possibly deleting cached entries.
+    // Because the `keys` entry is a point of contention among calls and we
+    // can only perform a single storage operation at the same time, we use
+    // a queue to ensure that all operations are done atomically.
+    return sbp('okTurtles.eventQueue/queueEvent', 'gi.db/files', () => {
+      // `_` prefix to avoid overlapping with the special `keys` key
+      return files.setItem('_' + cacheKey, blob).then(async v => {
```

Comment:
Instead of doing it this way, how about reversing it to simplify things?

Use the `cacheKey` as-is, and have a special `__keys` key that cannot be set using `'gi.db/files/save'`.

Reply:
It was based on this: <https://en.wikipedia.org/wiki/Hilbert%27s_paradox_of_the_Grand_Hotel>

I could do it the other way, but the special key still needs to be checked for and, throw an error, maybe?

Reply:
> I could do it the other way, but the special key still needs to be checked for and, throw an error, maybe?

Yeah 👍